docs/fsharp/style-guide/conventions.md

```diff
@@ -0,0 +1,648 @@
+
+```fsharp
+// This is bad!
+let tryReadAllText (path : string) =
+    try System.IO.File.ReadAllText path |> Ok
+    with e -> Error e.Message
+
+let r = tryReadAllText "path-to-file"
+match r with
+| Ok text -> ...
+| Error e ->
+    if e.Contains "uh oh, here we go again..." then ...
+    else ...
+```
+
+And placing the exception object itself in the `Error` constructor just forces you to properly deal with the exception type at the call site rather than in the function. Doing this effectively creates checked exceptions, which are notoriously unfun to deal with as a caller of an API.
+
+A good alternative to the above examples is to catch *specific* exceptions and return a meaningful value in the context of that exception. If you modify the `tryReadAllText` function as follows, `None` has more meaning:
+
+```fsharp
+let tryReadAllTextIfPresent (path : string) =
+    try System.IO.File.ReadAllText path |> Some
+    with :? FileNotFoundException -> None
+```
+
+Instead of functioning as a catch-all, this function will now properly handle the case when a file was not found and assign that meaning to a return. This return value can map to that error case, while not discarding any contextual information or forcing callers to deal with a case that may not be relevant at that point in the code.
+
+Types such as `Result<'Success, 'Error>` are appropriate for basic operations where they aren't nested, and F# optional types are perfect for representing when something could either return *something* or *nothing*. They are not a replacement for exceptions, though, and should not be used in an attempt to replace exceptions. Rather, they should be applied judiciously to address specific aspects of exception and error management policy in targeted ways.
+
+## Partial application and point-free programming
+
+F# supports partial application, and thus, various ways to program in a point-free style. This can be beneficial for code reuse within a module or the implementation of something, but it is generally not something to expose publicly. In general, point-free programming is not a virtue in and of itself, and can add a significant cognitive barrier for people who are not immersed in the style. Point-free programming in F# is basic for a well-trained mathematician, but can be difficult for people who are not familiar with lambda calculus.
+
+### Do not use partial application and currying in public APIs
+
+With little exception, the use of partial application in public APIs can be confusing for consumers. Usually, `let`-bound values in F# code are **values**, not **function values**. Mixing together values and function values can result in saving a small number of lines of code in exchange for quite a bit of cognitive overhead, especially if combined with operators such as `>>` to compose functions.
+
+### Consider the tooling implications for point-free programming
```

Comment:
the term "point-free programming" is new to me. Will your target audience know it?

Reply:
Yep! Well, mostly. Certainly FP programmers who come to F#.